contrib/symlink-githooks.sh

```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+remove_symlinks() {
+	for hook in githooks/*
+	do
+		bn=$(basename "$hook")
+		echo "Removing git hook symlink $bn"
+		rm ".git/hooks/$bn"
+	done
+	exit 0
+}
+
+add_symlinks() {
+	mkdir -p .git/hooks
+	for hook in githooks/*
+	do
+		bn=$(basename "$hook")
+		if [ ! -f ".git/hooks/$bn" ]; then
+			echo "Symlinking $bn..."
+			ln -s "../../githooks/$bn" ".git/hooks/$bn"
```

Comment:
I have a serious concern about the security implications. If a maintainers GH account gets compromised a malicious push could update the hooks to do something malicious thus compromising everyone else using them doing `git pull` and trying to make a contribution. I think we should copy them rather than symlink even though it makes updating them more annoying. If we want smooth updates at all costs maybe we should GPG-sign the hooks and instead copy a script that checks a hook and runs it.

Reply:
Agreed re "copying vs symlinking". It does make me nervous to have users auto-running stuff from the repo every time they invoke git.

The hooks should maybe hash themselves and compare to the hash in the repo to alert the user if the in-repo hooks have changed, and tell them how to re-do the copy.

Reply:
Good idea with hashing, just make sure to say "please review the hook diff" since if people blindly update we're at square one.

Reply:
Sure, we should say that. But even if people are blindly updating the situation is way better than if it were automatic. Since there would be a delay giving time for other maintainers to fix the problem and/or raise the alarm.

Reply:
>The hooks should maybe hash themselves and compare to the hash in the repo to alert the user if the in-repo hooks have changed, and tell them how to re-do the copy.

If the worry is that someone would push malicious hooks to the repo, then a symlinked hook would update during the next `git pull` and its hashed version would always match what was in the repo anyway.

Reply:
> its hashed version would always match what was in the repo anyway

Not if it's copied.

Reply:
Oh, I misunderstood.

I will change my PR (or leave it) based on whatever everyone decides.

Reply:
To clarify: I am in favor of copying instead of linking. So is @Kixunil. We should let @tcharding chime in as well.

Reply:
Perhaps we should split this work into two parts

1. This PR, and copy don't link
2. Another PR that adds a [post-merge](https://git-scm.com/docs/githooks#_post_merge) hook that compares the local hook (eg `.git/hooks/pre-commit`) with the one in `githooks/` - this is generally useful, turns out my current local hook is stale.

More about (2)

- Will need to be bullet proof, even for different dev set ups (eg don't hard code path `.git` vs `.bare`).
- We can just use `diff`, right? No need for the added complexity of hashing, AFAICT.

Reply:
(2) probably already exists on the internet, seems like a general problem that someone will have solved already.

Reply:
> turns out my current local hook is stale.

Quoting myself, this is a bit ironic since I was probably the one that patched the `pre-commit` hook that made my local one stale.

Reply:
I modified the PR to do a copy instead of linking. Also, I modified the part of the readme that suggested symlinking or setting the githooks config option to the repo's githooks. Lastly, I rebased on top the latest commits.